rust/src/lib.rs

```diff
@@ -64,25 +64,26 @@ pub mod objects;
 
 use std::collections::HashMap;
 
+use arrow_adbc_sys as adbc;
```

Comment:
In particular here - we don't want this dependency from the core ADBC to the bindings

Reply:
To be precise; the bindings here are generated items at build time based on `adbc.h`. The items are structs (e.g. `AdbcConnection`, `AdbcDatabase`), constants (e.g. `ADBC_VERSION_1_1_0`), functions (e.g. `AdbcDatabaseInit`) and type aliases (e.g. `AdbcStatusCode`).

We could choose to put some of the defintions behind a `ffi` feature, but I think always using the constants from the header is sensible, considering `adbc.h` the single source of truth for these.

Reply:
Good question. `bindgen` [uses](https://rust-lang.github.io/rust-bindgen/requirements.html#clang) `clang` to parse and analyze headers.

I think it's fine but others may disagree. I can revert this and keep copies of those definitions in Rust to prevent that dependency, and put the generated headers behind an `ffi` feature flag.

Or an alternative approach: check in the generated bindings (and regenerate when `adbc.h` changes).

Reply:
CC @wjones127 for opinions here, too

Reply:
I'll update this PR to check-in the generated bindings.

Reply:
> I know Go users often prefer to avoid CGO entirely if not required as it makes cross-compilation and static binaries harder

The only time I've had real issues in Rust like this was with `openblas-src`, and that was because of the Fortran compiler dependency. It's not uncommon in Rust to have some C dependencies; the most popular example is [openssl](https://crates.io/crates/openssl), which depends on `openssl-sys`.

I don't feel particularly strongly either way. I do suspect eventually someone might ask for a small pure-Rust version, and it seems reasonable that we build with that in mind. The amount of duplication doesn't seem that high to me.

I also do like the idea of having the bindings checked in somehow. It makes the code much easier to browse in GitHub.

Reply:
Hmm, but openssl is the canonical example of this: crates often have a feature to use rustls instead to avoid the native dependency, right?

Reply:
That's true, although I've never had issues with openssl that caused me to move to rustls. 🤷

Reply:
My main concern is locking us into a path that will be hard to back out of. I suppose if it's _only_ constant values then we can always copy them over later and break the dependency, but beyond that it may get painful if it turns out that pure-Rust is a major user concern.

Reply:
Updated the bindings crate to only generate when the `generate` feature is enabled (non-default), removing the `clang` dependency for the main crate. This then updates the checked-in bindings. CI checks if the bindings are up-to-date.

Reply:
I'd agree here: we don't want dependency from ADBC-Rust API to the bindings. It prevents pure-rust ADBC implementations.

General rule would be that ADBC-Rust API should not to contain anything C API specific.